Cmake: Add __ZEPHYR_SUPERVISOR__ macro for various files.

AdithyaBaglody opened the pull request:
Normally a syscall would check the current privilege level and then
decide to go to _impl_<syscall> directly or go through a
_handler_<syscall>.
__ZEPHYR_SUPERVISOR__ is a compiler optimization flag which will
make all the system calls from a driver/arch/kernel file to directly link
to the _impl_<syscall>. Thereby reducing the overhead of checking the
privileges.

AdithyaBaglody committed:
cmake: extensions: Added a new macro zephyr_library_ifdef

Checks if a certain Kconfig is enabled and then do the same
operation as zephyr_library.
If it finds that the Kconfig is not set it will not
create a new library and will stop processing the rest of
the CMakeFile.txt. Usefull in processing the driver Cmake
files.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
kernel: Cmake: Add __ZEPHYR_SUPERVISOR__ macro for kernel files.

Normally a syscall would check the current privilege level and then
decide to go to _impl_<syscall> directly or go through a
_handler_<syscall>.
__ZEPHYR_SUPERVISOR__ is a compiler optimization flag which will
make all the system calls from the kernel files directly link
to the _impl_<syscall>. Thereby reducing the overhead of checking the
privileges.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
arch: Cmake: Add __ZEPHYR_SUPERVISOR__ macro for arch files.

Normally a syscall would check the current privilege level and then
decide to go to _impl_<syscall> directly or go through a
_handler_<syscall>.
__ZEPHYR_SUPERVISOR__ is a compiler optimization flag which will
make all the system calls from the arch files directly link
to the _impl_<syscall>. Thereby reducing the overhead of checking the
privileges.

In the previous implementation all the source files would be compiled
by zephyr_source() rule. This means that zephyr_* is a catchall CMake
library for source files that can be built purely with the include
paths, defines, and other compiler flags that all zephyr source
files uses. This states that adding one extra compiler flag for only
one complete directory would fail.
This limitation can be overcome by using zephyr_libray* APIs. This
creates a library for the required directories and it also supports
directory level properties.
Hence we use zephyr_library* to create a new library with
macro _ZEPHYR_SUPERVISOR_ for the optimization.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
drivers: Cmake: Add __ZEPHYR_SUPERVISOR__ macro for driver files.

Normally a syscall would check the current privilege level and then
decide to go to _impl_<syscall> directly or go through a
_handler_<syscall>.
__ZEPHYR_SUPERVISOR__ is a compiler optimization flag which will
make all the system calls from the driver files directly link
to the _impl_<syscall>. Thereby reducing the overhead of checking the
privileges.

In the previous implementation all the source files would be compiled
by zephyr_source() rule. This means that zephyr_* is a catchall CMake
library for source files that can be built purely with the include
paths, defines, and other compiler flags that all zephyr source
files uses. This states that adding one extra compiler flag for only
one complete directory would fail.
This limitation can be overcome by using zephyr_libray* APIs. This
creates a library for the required directories and it also supports
directory level properties.
Hence we use zephyr_library* to create a new library with
macro _ZEPHYR_SUPERVISOR_ for the optimization.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

SebastianBoe commented on cmake/extensions.cmake:
```diff
@@ -389,6 +389,19 @@ macro(zephyr_library)
+# zephyr_library. If it finds that the Kconfig is not set it will not
+# create a new library and will stop processing the rest of
+# the CMakeFile.txt
+macro(zephyr_library_ifdef config)
```
This *_ifdef function should be located next to the other *_ifdef functions.

SebastianBoe replied:
Instead of 'config' should use the same naming convention as the other *_ifdef functions ('feature_toogle').

SebastianBoe commented on cmake/extensions.cmake:
```diff
@@ -389,6 +389,19 @@ macro(zephyr_library)
+      zephyr_library_named(${lib_name})
+      else()
+      return()
+    endif()
```
Also, should invoke zephyr_library instead of copying it's contents.

SebastianBoe commented on cmake/extensions.cmake:
```diff
@@ -389,6 +389,19 @@ macro(zephyr_library)
   zephyr_library_named(${lib_name})
 endmacro()
 
+# Checks if a Kconfig is enabled and then does the same operation as
```
The documentation should not mention Kconfig as there is nothing Kconfig-specific about this function.

SebastianBoe commented on arch/arc/core/CMakeLists.txt:
```diff
@@ -1,4 +1,6 @@
-zephyr_sources(
```
The commit message should explain why we are re-organizing from the zephyr_* API to the zephyr_library_* API.

(I know why, so it is not necessary to explain this to me)

SebastianBoe commented on drivers/adc/CMakeLists.txt:
```diff
@@ -1,8 +1,10 @@
-zephyr_sources_ifdef(CONFIG_ADC_QMSI_SS		  adc_qmsi_ss.c)
-zephyr_sources_ifdef(CONFIG_ADC_SAM_AFEC      adc_sam_afec.c)
-zephyr_sources_ifdef(CONFIG_ADC_TI_ADC108S102 adc_ti_adc108s102.c)
+zephyr_library_ifdef(CONFIG_ADC_HAS_DRIVER)
```
The need for _ifdef(ADC_HAS_DRIVER) is not clear to me.

I would think 

zephyr_library()

would be enough since drivers/CMakeLists.txt

is already doing:

add_subdirectory_if_kconfig(adc)

SebastianBoe replied:
"During the build someone would enable the CONFIG_ADC but before enabling this they don’t check if the given soc has a driver associated to it."

I still don't understand.

CONFIG_ADC_HAS_DRIVER means that the SoC supports the ADC driver.
CONFIG_ADC means that the user wishes to enable the ADC driver.

This CMakeLists.txt code is expressing that an ADC library should be enabled when an SoC supports ADC. But that is not correct, libraries should be enabled when the user has enabled
it. And enabling it should depend on the SoC supporting the driver. The dependency of having an enabled driver, and the driver being supported should be expressed in Kconfig.

So it should be something like this:

zephyr_library_ifdef(CONFIG_ADC)

config ADC
  depends on CONFIG_ADC_HAS_DRIVER

AdithyaBaglody replied:
In some test cases or samples they enable the driver but for the soc they write custom driver code or dummy code and they continue to use the Kconfig options available when then enable the driver. 
By doing this(as in the PR) I say, you can enable the driver but the soc specific code for the driver if not present in the drivers/{name} folder I don't compile anything inside it. But you can access any of the Kconfigs which get enabled when by saying CONFIG_ADC=y.
 
I think the method I used works without breaking any samples/testcases . This way its backward compatible. I feel more secure doing it this way(as in the PR).

SebastianBoe replied:
"In some test cases or samples they enable the driver but for the soc they write custom driver code or dummy code and they continue to use the Kconfig options available when then enable the driver."

What you are describing is an invalid use of the configuration system. We cannot write (invalid) build scripts to make invalid configurations build. Instead, the invalid use must be rectified.

One problem with the scheme that you are proposing is that we can no longer have the configuration system enforce valid configurations. E.g. if there is no dependency
 of enabling ADC on ADC being supported, we can not prevent users from creating invalid configurations. Enforcing valid configurations is one of the primary objectives of the 
configuration system so it is not something we can drop.

It is okay to break samples/tests when they rely on invalid behaviour.

AdithyaBaglody replied:
@SebastianBoe I understand the concern here. I tried doing this for FLASH and ADC. there are too many dependencies and it is not straight forward and is very tedious. Also this would not be in the scope of this PR. 
So the best thing I can see is either we continue with this approach or we do not include _supervisor_ macro for drivers. I am inclined to not include the macro.
If and when we need the optimization for the drivers, we can take this up then.
(@andrewboie  what do you think?)

SebastianBoe replied:
Fair enough, you have a point that a failure in one subsystem should not halt progress in another.

But if we approve this PR then this means we are in agreement that:

We are writing Kconfig in a way that does not allow enforcement of valid configurations.

We do not know how to write Kconfig in a way that allows enforcement of valid configurations,
and therefore must continue to write invalid Kconfig until we can give guidelines that can be enforced through code-review.

nashif replied:
this needs more thought, I dislike this change adding all of those Kconfig dependencies and hidden variables. Will sync with @AdithyaBaglody to understand the issue better and possibly come up with a better solution.

SebastianBoe reviewed (changes requested):
Review comments are inline.

AdithyaBaglody committed:
tests: Kconfig: Added a new Kconfig for qemu_x86

The qemu doesn't have any flash drivers present in the
drivers/flash folder.
When building the tests/subsys/fs/nffs_fs_api the qemu picks
up a local file to emulate the drivers. This setup was causing the
failures in mapping required dependencies. This patch creates a
new kconfig which can be used to test out the flash drivers.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
linker-defs: Increase the number of kernel objects

Required to compile large samples/tests.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

tyrantlizard77 reviewed (approved):
The library split up changes I do not completely understand but I'm deferring to @SebastianBoe opinion.

Did we resolve the increased footprint issues you were dealing with?

Would like @nashif blessing on this approach before it goes in since this is a nontrivial change to our Kconfig ontology.

andrewboie commented:
This needs a rebase.
Also need approval from @SebastianBoe

AdithyaBaglody committed:
kernel: thread: Move out the function _thread_entry() to lib

The _thread_entry() is not really a part of the kernel but a part of
the zephyr's C runtime support library. Hence moving just the
function to lib/thread_entry.c

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

nashif reviewed (changes requested):
need to reconsider the driver Kconfig change, will look for possible alternatives and come back.

nashif commented:
@AdithyaBaglody can you give an example of samples/tests that would fail if we do not apply the HAS_DRIVER logic?

nashif commented:
@SebastianBoe see comment above

SebastianBoe commented:
@nashif : The main problem with adding a dummy c file is that we will no longer be able to detect at build-time that the user has accidentally enabled a driver without enabling a driver that implements it.

This misconfiguration has been common and is why we have this error-feedback:

```
  # Verify that all (non-imported) libraries have source
  # files. Libraries without source files are not supported because
  # they are an indication that something has been misconfigured.
  get_target_property(lib_imported ${zephyr_lib} IMPORTED)
  get_target_property(lib_sources  ${zephyr_lib} SOURCES)
  if(lib_sources STREQUAL lib_sources-NOTFOUND
      AND (NOT (${zephyr_lib} STREQUAL app))
      AND (NOT lib_imported)
      )
    # app is not checked because it's sources are added to it after
    # this CMakeLists.txt file has been processed
    message(FATAL_ERROR "\
The Zephyr library '${zephyr_lib}' was created without source files. \
Empty (non-imported) libraries are not supported. \
Either make sure that the library has the sources it should have, \
or make sure it is not created when it has no source files.")
  endif()

```

Therefore I think it would be better for the user if we "fix the tests and add platform filtering and fix tests".

nashif commented:
> Therefore I think it would be better for the user if we "fix the tests and add platform filtering and fix tests".

yeah, i need to look into what this would take.

AdithyaBaglody committed:
tests: drivers: build_all: Ensure correct builds.

Ensure that sanitycheck runs this sample only on boards with
right drivers support. This test was picking all targets and
executing on each of them .This eliminates incorrect builds.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: subsys: fs: Enable proper configuration for qemu_x86.

The sample was running with incorrect kconfigs.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented:
@nashif I have done those changes we discussed. Can you please have a look.

nashif commented on samples/grove/light/sample.yaml:
```diff
@@ -5,5 +5,6 @@ tests:
     platform_whitelist: arduino_101_sss quark_d2000_crb
       arduino_due
     tags: drivers sensor grove light
+    depends_on: adc
```
duplicate

nashif commented on tests/power/power_states/testcase.yaml:
```diff
@@ -1,12 +1,12 @@
   test:
     filter: (CONFIG_SOC_QUARK_SE_C1000 or CONFIG_SOC_QUARK_SE_C1000_SS)
-    platform_exclude: tinytile
+    platform_whitelist: quark_se_c1000 quark_se_c1000_ss
```
this should already be addressed with `depends_on: rtc `below

nashif commented on tests/power/power_states/testcase.yaml:
```diff
@@ -1,12 +1,12 @@
     extra_args: CONF_FILE="prj_socwatch.conf"
     filter: CONFIG_SOC_QUARK_SE_C1000
-    platform_exclude: tinytile
+    platform_whitelist: quark_se_c1000
```
ditto

nashif commented on samples/boards/quark_se_c1000/power_mgr/sample.yaml:
```diff
@@ -3,5 +3,6 @@ sample:
 tests:
   test:
     filter: (CONFIG_SOC_QUARK_SE_C1000 or CONFIG_SOC_QUARK_SE_C1000_SS)
+    platform_whitelist: quark_se_c1000 quark_se_c1000_ss
```
addressed with dependency on rtc below

nashif commented on drivers/flash/Kconfig:
```diff
@@ -10,13 +10,13 @@
 #
 config FLASH_HAS_DRIVER_ENABLED
-	bool
+	bool "Has enabled the driver"
```
_HAS_ options need to remain hidden, this is not how we should solve this issue

nashif commented on drivers/flash/Kconfig:
```diff
@@ -10,13 +10,13 @@
 
 config FLASH_HAS_PAGE_LAYOUT
-	bool
+	bool "Has page layout"
```
ditto

AdithyaBaglody commented:
@nashif fixed the review comments.

nashif commented:
recheck

SebastianBoe commented on cmake/extensions.cmake:
```diff
@@ -905,6 +905,18 @@ macro(list_append_ifdef feature_toggle list)
+    if(${${feature_toggle}})
+      zephyr_library()
+      else()
+      return()
```
Issue created: #7565
